### PR TITLE
[SPARK-48493][PYTHON] Enhance Python Datasource Reader with direct Arrow Batch support for improved performance

### PR DESCRIPTION
**Pull request by LucaCanali**
### What changes were proposed in this pull request?
This pull request proposes enhancing the Python Datasource Reader by adding an option to yield Arrow batches directly. This change aims to significantly improve performance compared to the existing approach of using tuples or Rows. The implementation takes advantage of the existing work with MapInArrow (referenced in SPARK-46253).

### Why are the changes needed?
The changes are needed to address performance issues in the Python Datasource Reader. The current method of sending data as tuples or Rows is inefficient, leading to slower data processing times. By allowing the Datasource Reader to yield Arrow batches directly, we can use the more efficient Arrow format, significantly speeding up data processing. Tests have shown this approach to be up to 8x faster (in a preliminary test with a High Energy Physics Datasource reader for the ROOT data format), particularly benefiting use cases involving large datasets.

### Does this PR introduce _any_ user-facing change?
Yes, this PR introduces a user-facing change by adding an option to the Python Datasource Reader that allows users to yield Arrow batches directly.

### How was this patch tested?
A new test was added to the Python Datasource test suite. Additionally, it was manually tested using a custom Python datasource for performance testing.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: python/pyspark/sql/tests/test_python_datasource.py [hunk 1]
```diff
@@ -373,6 +373,57 @@ def test_case_insensitive_dict(self):
...
+            def name(cls):
+                return "arrowbatch"
+
+            def schema(self):
+                return "key int, value string"
+
+            def reader(self, schema: StructType):
+                return ArrowBatchDataSourceReader(schema, self.options)
+
+        class ArrowBatchDataSourceReader(DataSourceReader):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_python_datasource.py:394]: We should probably change the type hint of `DataSourceReader.read`.

**Review by allisonwang-db — COMMENTED** [review]
I think this is a great idea, but instead of adding a config `readerYieldsArrowBatches` in the data source `option`, we can add support in the API layer to directly yield arrow batches for data source read:
```python
@abstractmethod
def read(self, partition: InputPartition) -> Union[Iterator[Tuple], Iterator[pa.RecordBatch]]:
```
and in the `records_to_arrow_batches` we can check if the `output_iter` is an iterator of arrow batches and directly return that.

File: python/pyspark/sql/datasource.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
 #
 from abc import ABC, abstractmethod
 from collections import UserDict
+import pyarrow as pa
```
Comment (allisonwang-db) [inline python/pyspark/sql/datasource.py:19]: Hmm I don't think we can directly import pyarrow here unless we make pyarrow the one of the default dependency for pyspark. cc @HyukjinKwon WDYT?

File: python/pyspark/sql/worker/plan_data_source_read.py [hunk 3]
```diff
@@ -54,17 +54,31 @@
 
 
 def records_to_arrow_batches(
```
Comment (allisonwang-db) [inline python/pyspark/sql/worker/plan_data_source_read.py:56]: This method is also used by streaming reader. cc @chaoqin-li1123 do you think we should support arrow batch output for streaming reader as well?

File: python/pyspark/sql/worker/plan_data_source_read.py [hunk 4]
```diff
@@ -54,17 +54,31 @@
...
     """
-    Convert an iterator of Python tuples to an iterator of pyarrow record batches.
+    First check if the iterator yields pyarrow record batched, if so, yield them directly.
+    Otherwise, convert an iterator of Python tuples to an iterator of pyarrow record batches.
+    For each Python tuple, check the types of each field and append it to the records batch.
+    """
 
-    For each python tuple, check the types of each field and append it to the records batch.
+    # Create a copy of the iterator using itertools.tee
+    output_iter, output_iter_copy = tee(output_iter)
```
Comment (allisonwang-db) [inline python/pyspark/sql/worker/plan_data_source_read.py:69]: It might be inefficient to create a copy of the entire `output_iter` here.

File: python/pyspark/sql/worker/plan_data_source_read.py [hunk 5]
```diff
@@ -54,17 +54,31 @@
...
 
-    """
+    try:
+        # Check if the first element is of type pa.RecordBatch
+        # In that case, yield all elements and return
+        first_element = next(output_iter_copy)
+        if isinstance(first_element, pa.RecordBatch):
+            yield first_element
+            for element in output_iter_copy:
+                yield element
```
Comment (allisonwang-db) [inline python/pyspark/sql/worker/plan_data_source_read.py:78]: I think we need to check if the schema of the arrow batch  (# of output columns, data types, etc)  matches the expected schema of the data source (similar to the checks below for iterator[row])

File: python/pyspark/sql/tests/test_python_datasource.py [hunk 6]
```diff
@@ -373,6 +373,53 @@ def test_case_insensitive_dict(self):
...
+            def __init__(self, schema, options):
+                self.schema: str = schema
+                self.options = options
+
+            def read(self, partition):
+                # Create Arrow Record Batch
+                keys = pa.array([1, 2, 3, 4, 5], type=pa.int32())
+                values = pa.array(["one", "two", "three", "four", "five"], type=pa.string())
+                schema = pa.schema([("key", pa.int32()), ("value", pa.string())])
```
Comment (allisonwang-db) [inline python/pyspark/sql/tests/test_python_datasource.py:403]: Can we see what error message we will get if the schema here is different from the data source schema?

**Review by allisonwang-db — COMMENTED** [review]
Left some comments. I really like this idea!

File: python/pyspark/sql/worker/plan_data_source_read.py (hunk 3 above)
Comment (chaoqin-li1123) [reply python/pyspark/sql/worker/plan_data_source_read.py:56]: I don't have a strong opinion here. For API consistency I prefer to make batch and streaming data source has the same return type.

File: python/pyspark/sql/datasource.py (hunk 2 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/datasource.py:19]: yeah let's don't import here

Comment (HyukjinKwon) [reply python/pyspark/sql/datasource.py:19]: You can do with `if typing.TYPE_CHECKING` and import pyarrow as pa, then type it as `"pa.RecordBatch"` below.

File: python/pyspark/sql/worker/plan_data_source_read.py (hunk 4 above)
Comment (LucaCanali) [reply python/pyspark/sql/worker/plan_data_source_read.py:84]: Good point. Any tips?

File: python/pyspark/sql/worker/plan_data_source_read.py (hunk 5 above)
Comment (LucaCanali) [reply python/pyspark/sql/worker/plan_data_source_read.py:114]: Indeed. I have added a check on # of output columns and column names for the moment. Probably data types should be done too.

**Comment by HyukjinKwon** [comment]
I will defer to @allisonwang-db

File: python/pyspark/sql/datasource.py [hunk 7]
```diff
@@ -332,8 +331,13 @@ def partitions(self) -> Sequence[InputPartition]:
             message_parameters={"feature": "partitions"},
         )
 
+    if TYPE_CHECKING:
+        from pyarrow import RecordBatch
+    else:
+        RecordBatch = None
```
Comment (allisonwang-db) [inline python/pyspark/sql/datasource.py:337]: How about let's move this to the beginning of this file? Then we don't have to import this multiple times.
```
if TYPE_CHECKING:
    from pyarrow import RecordBatch
    from pyspark.sql.session import SparkSession
```

File: python/run-tests.py [hunk 8]
```diff
@@ -207,9 +207,9 @@ def run_individual_python_test(target_dir, test_name, pyspark_python, keep_test_
 
 
 def get_default_python_executables():
-    python_execs = [x for x in ["python3.9", "pypy3"] if which(x)]
+    python_execs = [x for x in ["python3", "pypy3"] if which(x)]
 
-    if "python3.9" not in python_execs:
```
Comment (allisonwang-db) [inline python/run-tests.py:212]: This change seems unrelated?

**Review by allisonwang-db — COMMENTED** [review]
Looks good overall!

**Comment by LucaCanali** [comment]
Hi @allisonwang-db I hope you're doing well! When you have a moment, could you please review the latest changes? Your feedback would be greatly appreciated.

**Review by allisonwang-db — APPROVED** [review]
@LucaCanali thanks for the ping! This PR looks good overall. As a follow up, can we add some documentations and examples in the datasource user guide here? `python/docs/source/user_guide/sql/python_data_source.rst`

File: python/docs/source/user_guide/sql/python_data_source.rst [hunk 9]
```diff
@@ -452,3 +452,67 @@ We can also use the same data source in streaming reader and writer
 .. code-block:: python
 
     query = spark.readStream.format("fake").load().writeStream.format("fake").start("/output_path")
+
+Python Datasource Reader with direct Arrow Batch support for improved performance
+-----------------------------------------------------------------------------------------
+The Python Datasource Reader now supports direct yielding of Arrow Batches, which can significantly improve data processing performance. By using the efficient Arrow format,
```
Comment (allisonwang-db) [inline python/docs/source/user_guide/sql/python_data_source.rst:458]: 🎉

**Review by allisonwang-db — APPROVED** [review]
Awesome!

**Comment by allisonwang-db** [comment]
@LucaCanali please re-trigger the failed tests (they seems unrelated to this change)

**Comment by LucaCanali** [comment]
Somehow the test org.apache.spark.sql.streaming.ClientStreamingQuerySuite keeps failing, although it seems unrelated to this change?

**Comment by HyukjinKwon** [comment]
Seems unrelated but mind syncing the branch to the latest master and retrigger the test? The change here isn't minor so let;s make sure the tests pass.

**Comment by LucaCanali** [comment]
I guess this should be good to go now. Any further comments?

**Comment by allisonwang-db** [comment]
Thanks, merging to master.

**Comment by LucaCanali** [comment]
Thank you @allisonwang-db and @HyukjinKwon !
